teps/0143-concise-parameters-results.md

```diff
@@ -0,0 +1,848 @@
+            script: |
+              BASE_URL=$(echo "$(params.releaseBucket)/previous/$(params.versionTag)")
+              # If the bucket is in the gs:// return the corresponding public https URL
+              BASE_URL=$(echo ${BASE_URL} | sed 's,gs://,https://storage.googleapis.com/,g')
+              echo "${BASE_URL}/release.yaml" > $(results.release.path)
+              echo "${BASE_URL}/release.notag.yaml" > $(results.release-no-tag.path)
+```
+
+### Goals
+
+- Provide backward-compatible concise syntax for `Parameters` and `Results`.
+
+### Prior Art and Related Work
+
+There have been efforts in Tekton to make the API simple and concise.
+
+#### Mapping Workspaces
+
+Tekton auto-maps `Workspaces` from `Pipelines` to `PipelineTasks` when the names match, per [TEP-0108](0108-mapping-workspaces.md) to reduce the verbosity in mapping `Workspaces`.
+
+#### Propagated Workspaces and Parameters
+
+Tekton propagates `Workspaces` and `Parameters` in embedded specifications, per [TEP-0111](0111-propagating-workspaces.md) and [TEP-0107](0107-propagating-parameters.md) to reduce the verbosity of repeatedly defining `Workspaces` and `Parameters`.
+
+#### Github Actions
+
+Github Actions uses maps for `Inputs` and `Outputs`. Take the [Cloud Functions Deploy][cfd] for example:
+
+```yaml
+name: Cloud Functions Deploy
+description: Use this action to deploy your function or update an existing Cloud Function.
+inputs:
+  name:
+    description: Name of the Cloud Function.
+    required: true
+  description:
+    description: Description for the Cloud Function.
+    required: false
+  runtime:
+    description: Runtime to use for the function.
+    required: true
+outputs:
+  url:
+    description: The URL of your Cloud Function. Only available with HTTP Trigger.
+  id:
+    description: Full resource name of the Cloud Function, of the format 'projects/p/locations/l/functions/f'.
+  status:
+    description: Status of the Cloud Function deployment.
+  version:
+    description: Version of the Cloud Function deployment.
+  runtime:
+    description: Runtime of the Cloud Function deployment.
+runs:
+  using: 'node16'
+  main: 'dist/index.js'
+```
+
+## Proposal
+
+We propose using maps for `Params` and `Results` instead of name and value pairs.
+
+### Syntax
+
+#### Parameters
+
+At authoring time, users declare string, array and object `Parameters` for `Tasks` and `Pipelines`. It is tedious to specify these fields because they are defined as sub-objects. We propose using maps instead of sub-objects to declare `Parameters`. The new syntax of `Params` will go under a new api field `Inputs.Params`. This is because we cannot modify the existing `Params` field because of our [Go compatibility policy][cp].
+
+```yaml
+# before
+params:
+  - name: hello
+    type: string
+  - name: environments
+    type: array
+  - name: git
+    type: object
+    properties:
+      url: {}
+      commit: {}
+
+# after
+inputs:
+  params:
+    hello:
```

Comment:
This is fine and safe. It will be validated easily during the creation. It will be put to
```
map[string]struct.
```